rfd/0153-resource-guidelines.md

```diff
@@ -0,0 +1,469 @@
+---
+authors: Tim Ross (tim.ross@goteleport.com)
+state: draft
+---
+
+# RFD 153 - Resource Guidelines
+
+## Required approvers
+
+- Engineering: `@zmb3`
+
+## What
+
+Guidelines for creating backend resources and APIs to interact with them.
+
+## Why
+
+To date resources follow slightly different patterns and have inconsistent APIs for creating, updating and deleting
+them, see [#29234](https://github.com/gravitational/teleport/issues/29234). In addition to reducing confusion and
+fostering a better developer experience, this will also remove complexity from the terraform provider and teleport
+operator that have to accommodate for every subtle API and resource difference that exist today.
+
+## Details
+
+### Defining a resource
+
+A resource at minimum MUST include a kind, version, `teleport.header.v1.Metadata` and a specification message with any
+additional parameters needed to represent the resource. The kind, version and metadata helps to identify the type of
+resource, when it should expire, the version of the resource, etc. This is particularly important for ensuring that all
+resources contain the same shape and that tools like `tctl` are able to leverage `services.UnknownResource` to identify
+the resource and act appropriately.
+
+```protobuf
+import "teleport/header/v1/resourceheader.proto";
+
+// Foo is a resource that does foo.
+message Foo {
+  // The kind of resource represented.
+  string kind = 1;
+  // An optional subkind to differentiate variations of the same kind.
+  string sub_kind = 2;
+  // The version of the resource being represented.
+  string version = 3;
```

Comment:
Maybe note somewhere that `CheckAndSetDefaults` should fail on unknown version values, even for the initial release of the resource (its been a common error in the past to not start caring about `Version` until we need to bump it... then having to deal with older clients silently accepting unrecognized versions and doing the wrong thing with them).

Reply:
I think that's fine, thanks.